modules/game_market/classes/t_history.lua

```diff
@@ -0,0 +1,286 @@
+MarketHistory = {}
+MarketHistory.__index = MarketHistory
+
+local onTopListValueChange = nil
+local onBottomListValueChange = nil
+
+local topListMin = 0
+local topListMax = 0
+
+local bottomListMin = 0
+local bottomListMax = 0
+
+local topListFitItems = 0
+local bottomListFitItems = 0
+
+local labelSize = 16
+local historyOfferPool = 14
+
+local topListPool = {}
+local bottomListPool = {}
+
+local topListData = {}
+local bottomListData = {}
+
+function MarketHistory.onTopListValueChange(scroll, value, delta)
+  local startLabel = math.max(topListMin, value)
+  local endLabel = startLabel + topListFitItems - 1
+
+  if endLabel > topListMax then
+    endLabel = topListMax
+    startLabel = endLabel - topListFitItems + 1
+  end
+
+  for i, widget in ipairs(topListPool) do
+    local index = value > 0 and (startLabel + i - 1) or (startLabel + i)
+    local data = topListData[index]
+	if not data then
+		widget:setVisible(false)
+		goto continue
+	end
+
+	local color = i % 2 == 0 and '#484848' or '#414141'
+	widget:setId(color)
+	widget.actionId = i
+	widget:setBackgroundColor(color)
+		widget:setColor('#c0c0c0')
+		widget.amount:setText(data.amount)
+		widget.name:setText(g_things.getThingType(data.itemId, ThingCategoryItem):getMarketData().name)
+		widget.endAt:setText(os.date("%Y-%m-%d, %H:%M:%S", data.timestamp))
+		widget.status:setText(MarketSellStatus[data.state])
+		widget.piecePrice:setColor("#c0c0c0")
+		widget.totalPrice:setColor("#c0c0c0")
+		widget.name:setColor("#c0c0c0")
+		widget.amount:setColor("#c0c0c0")
+		widget.endAt:setColor("#c0c0c0")
+		widget.status:setColor("#c0c0c0")
+
+		if data.itemTier > 0 then
+			widget.name:setText(widget.name:getText() .. " (Tier " .. data.itemTier .. ")")
+		end
+
+		local totalPrice = data.price * data.amount
+		local unitPrice = data.price
+		widget.piecePrice:setText(convertLongGold(unitPrice))
+		widget.totalPrice:setText(convertLongGold(totalPrice))
+		if totalPrice > 99999999 then
+			widget.totalPrice:setTooltip(comma_value(totalPrice))
+		end
+
+	if unitPrice > 99999999 then
+		widget.piecePrice:setTooltip(comma_value(unitPrice))
+	end
+
+	::continue::
```

Comment:
```suggestion
	::continue::
```

this breaks lua 5.1, never use continue

Reply:
Got it, will fix that.